File: program/arkanoid_component/src/gamestates/MainGameState.cpp
/*
 * Copyright (C) 2014 by Francois Coulombe
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:

 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.

 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */


#include "MainGameState.h"


#include <applayer/Actor.h>
#include <applayer/GameStateManager.h>
#include <applayer/SpriteComponent.h>
#include <applayer/Sprite.h>
#include <gcl/Config.h>
#include <gcl/Exception.h>
#include <gcl/Time.h>
#include <input/Input.h>

using namespace GCL;
using namespace Arkanoid;

static const size_t NUM_ROW 	= 5;
static const size_t NUM_COL 	= 8;
static const size_t BORDER_SIZE = 0;
static const size_t BASE_X_RESOLUTION = 1024;
static const size_t BASE_Y_RESOLUTION = 768;
static const size_t VIEWPORT_WIDTH =  Config::Instance().GetInt("DEFAULT_VIEWPORT_WIDTH") ;
static const size_t VIEWPORT_HEIGHT =  Config::Instance().GetInt("DEFAULT_VIEWPORT_HEIGHT") ;
static const Real X_RATIO = (Real)VIEWPORT_WIDTH/ (Real)BASE_X_RESOLUTION;
static const Real Y_RATIO = (Real)VIEWPORT_HEIGHT / (Real)BASE_Y_RESOLUTION;

static const WorldPoint3 START_POSITION(640.0*X_RATIO, 700.0*Y_RATIO, 0.0);
static const WorldPoint3 BALL_INITIAL_VELOCITY(-6.0, -3.0, 0.0);

MainGameState::~MainGameState()
{
	for (size_t i=0; i<blocks.size(); ++i)
	{
		delete blocks[i];
	}
	blocks.clear();

	delete paddle;
	delete ball;
}

MainGameState::MainGameState()
:GameState("MainGame")
{
	std::cout << "EnterMainGameState"<<std::endl;
	for (size_t i=0; i<NUM_ROW*NUM_COL; ++i)
	{
		Actor *block = new Actor("Block", "Block");
		blocks.push_back(block);
		Sprite & sprite = *static_cast<SpriteComponent*>(block->GetComponent("SpriteComponent"))->GetSprite();

		size_t blockWidth = sprite.GetWidth();
		size_t blockHeight = sprite.GetHeight();
		Real blockHalfWidth = blockWidth/2.0;
		Real blockHalfHeight = blockHeight/2.0;

		size_t spaceX = (VIEWPORT_WIDTH)/NUM_COL;
		size_t spaceY = blockHeight;

		Real x = (int(i%NUM_COL)*spaceX) + BORDER_SIZE + blockHalfWidth;
		Real y = (int(i/NUM_COL)*spaceY) + BORDER_SIZE + blockHalfHeight;

		//
		WorldPoint3 position(x,y, 0.0);
		block->SetPosition(position);
		std::cout << position << std::endl;
	}
	paddle = new Actor("Paddle", "Paddle");
	ball = new Actor("Ball", "Ball");
	paddle->SetPosition(START_POSITION);

	ball->SetPosition(START_POSITION+WorldPoint3(100,0,0));
}
bool MainGameState::Update(Real dt)
{
	//std::cout << "MainGameState";
	for (size_t i=0; i<NUM_ROW*NUM_COL; ++i)
	{
		blocks[i]->Update(dt);
	}
	paddle->Update(dt);
	ball->Update(dt);
	GameState::Update(dt);

	return true;
}

File: program/arkanoid_component/data/Script/MainGameState.lua
function CreateBlocks()
  for i=1, NUM_ROW*NUM_COL do
    local blockName = "Block" .. i
    kinevox.Log("Creating: " .. blockName)
    kinevox.CreateActor(blockName, "Block")
    local block = kinevox.GetActor(blockName)
    
    local blockWidth = 64
    local blockHeight = 32
    local blockHalfWidth = blockWidth/2.0
    local blockHalfHeight = blockHeight/2.0

    local spaceX = (VIEWPORT_WIDTH)/NUM_COL
    local spaceY = blockHeight

    local x = ((i%NUM_COL)*spaceX) + BORDER_SIZE + blockHalfWidth
    local y = ((i/NUM_COL)*spaceY) + BORDER_SIZE + blockHalfHeight

    block:SetPosition(x, y, 0.0)
  end  
end

function CreatePaddle()
  kinevox.CreateActor("Paddle", "Paddle")
  paddle = kinevox.GetActor("Paddle")
  paddle:SetPosition(START_POSITION[1], START_POSITION[2], START_POSITION[3])
end

function CreateBall()
  kinevox.CreateActor("Ball", "Ball")
  ball = kinevox.GetActor("Ball")
  ball:SetPosition(START_POSITION[1]+100, START_POSITION[2], START_POSITION[3])
end

function Initialize(self)
  kinevox.Log("Main Game State!")
  NUM_ROW   = 5;
  NUM_COL   = 8;
  BORDER_SIZE = 0;
  BASE_X_RESOLUTION = 1024;
  BASE_Y_RESOLUTION = 768;
  SCREEN_SIZE = kinevox.GetScreenSize()
  VIEWPORT_WIDTH =  SCREEN_SIZE[0]
  VIEWPORT_HEIGHT =  SCREEN_SIZE[1]
  X_RATIO = VIEWPORT_WIDTH / BASE_X_RESOLUTION
  Y_RATIO = VIEWPORT_HEIGHT / BASE_Y_RESOLUTION
  kinevox.Log("ratio! "..X_RATIO.." "..Y_RATIO)
  
  
  START_POSITION = {640.0*X_RATIO, 700.0*Y_RATIO, 0.0}
  CreateBlocks()
  CreatePaddle()
  CreateBall()
end

function Logic(self, dt)

end

function Terminate(self)
  kinevox.DestroyActor("Ball")
  kinevox.DestroyActor("Paddle")

  for i=1, NUM_ROW*NUM_COL do
    kinevox.DestroyActor("Block" .. i)
  end
end